### PR TITLE
Add a flag to thriftbreak to output results as JSON.

### PR DESCRIPTION
**Pull request by shirchen**
Adding `--json` flag to thriftbreak to output results as JSON. This enables integration with tools like Phabricator.

Ref GO-938

### CODE REVIEW

File: cmd/thriftbreak/main.go
Comment (abhinav) [inline cmd/thriftbreak/main.go:62]: This variable declared outside the loop seems unnecessary.
It's only needed inside the loop.

File: cmd/thriftbreak/main.go
Comment (abhinav) [inline cmd/thriftbreak/main.go:70]: This if-else over here is a bit smelly.

Consider resolving the way output is processed once outside the loop (possibly into a `func(Lint) error`)
and calling that from within the loop.

So maybe something like the following would work:

```go
func readableOutput(w io.Writer) func(Lint) error
func jsonOutput(w io.Writer) func(Lint) error
```

Usage:

```go
var write func(Lint) error
if *json { write = jsonOutput(os.Stdout) }
else { write = readableOutput(os.Stdout) }

for _, l := range lints {
  if err := write(l); err != nil { .. }
}
```

---

For the JSON output, if you use [json.Encoder](https://pkg.go.dev/encoding/json#Encoder), the newline will be included for you:

```
  enc := json.NewEncoder(w)
  return func(l Lint) error {
    // Encode adds a trailing newline.
    if err := enc.Encode(l); err != nil {
        return fmt.Errorf(..)
    }
    return nil
  }
```

---

Lastly, I'm using `func(Lint) error` as the output consumer,
but I'll leave it up to you if that's okay,
or if you'd prefer something like a `Sink interface{ WriteLint(Lint) error }` type.

File: cmd/thriftbreak/main.go
Comment (abhinav) [inline cmd/thriftbreak/main.go:42]: minor: If we're adding a JSON flag, we should maybe mention what fields are included in the JSON object.

    "output as a list of newline-delimited JSON objects with the following fields: File and Message"

(reformat and reword for readability to your preference.)

File: cmd/thriftbreak/main_test.go
Comment (abhinav) [inline cmd/thriftbreak/main_test.go:41]: Naming this type is unnecessary. For table tests, we typically don't name the test case struct unless we need to declare a separate `func` that accepts it.

```
tests := []struct {
  // fields
}{
  // test cases
}
```

File: cmd/thriftbreak/main_test.go
Comment (abhinav) [inline cmd/thriftbreak/main_test.go:91]: Seeing as these bits are constant, consider putting this outside the loop, above the test cases.

```
from := ...
to := ...
remove := ...

tests := []struct {
 ...
}{
  ..
}

for _, tt := range tests {
  t.Run(..., func(t *testing.T) {
    // ...
    breaktest.CreateRepoAndCommit(..)
  })
}
```

File: cmd/thriftbreak/main_test.go
Comment (abhinav) [inline cmd/thriftbreak/main_test.go:100]: I don't remember if we previously discussed this.
stdout hijacking is undesirable.
It is preferable for `run` to parameterize stdout as an io.Writer parameter and write to that.

That will also significantly simplify this test case by obviating the need for the TempFile. We'll be able to pass a bytes.Buffer as the stdout io.Writer and verify output.

File: cmd/thriftbreak/main.go
Comment (sywhang) [inline cmd/thriftbreak/main.go:64]: If we do this, aren't we appending many JSONs, separated by a newline? Sounds like that wouldn't be a valid JSON format if you piped the output - is that what we want or should we output one valid JSON-formatted string instead?

File: cmd/thriftbreak/main_test.go
Comment (shirchen) [reply cmd/thriftbreak/main_test.go:100]: Sounds good. Yup, noted in GO-892

File: cmd/thriftbreak/main.go
Comment (abhinav) [reply cmd/thriftbreak/main.go:64]: I think the typical behavior for something like this is newline delimited JSON objects. These can be fed directly into jq.

Comment (abhinav) [reply cmd/thriftbreak/main.go:64]: (A single JSON string can also be fed into jq, but there's explicit support for ndjson in jq.)

Comment (sywhang) [reply cmd/thriftbreak/main.go:64]: i see. nvm then :)

File: cmd/thriftbreak/main.go
Comment (sywhang) [inline cmd/thriftbreak/main.go:58]: Do we care about failure cases being tested here?

File: cmd/thriftbreak/main_test.go
Comment (sywhang) [inline cmd/thriftbreak/main_test.go:96]: stdout hijacking is probably unnecessary now that you have jsonOutput and readableOutput but I assume you'll be doing that in a follow-up PR?

File: cmd/thriftbreak/main.go
Comment (shirchen) [reply cmd/thriftbreak/main.go:59]: yeah, im not sure how to make serialization of a struct that we own fail. fields are strings too so i couldnt figure out how to make a test for catching an error.

File: cmd/thriftbreak/main.go
Comment (shirchen) [reply cmd/thriftbreak/main.go:45]: I dont think it does, so leaving that here.

**Review by sywhang — APPROVED** [review]
:shipit:

File: cmd/thriftbreak/main_test.go
Comment (abhinav) [reply cmd/thriftbreak/main_test.go:99]: I think the hijacking is still needed if while the code is still calling `run()` because run does not yet parameterize stdout. @shirchen plans to fix that in GO-892.

The unit tests for json/readableOutput don't need stdout hijacking, though, that's true.

File: cmd/thriftbreak/main_test.go
Comment (abhinav) [inline cmd/thriftbreak/main_test.go:108]: So two things:

- the contents of gitDirArg don't vary
- the flag doesn't need to use the "-C=foo" form to be picked up

The suggestion from the first one would have been to just do:

```
run([]string{"-C="+tmpDir, extraCmd})
```

But also because you don't need to use `=`, tmpDir can just be the next arg, you can also do the following, which eliminates the string append too:

```suggestion
			err = run([]string{"-C", tmpDir, extraCmd})
```

File: cmd/thriftbreak/main_test.go
Comment (abhinav) [inline cmd/thriftbreak/main_test.go:107]: It's unclear why these variables are being declared here.
What scoping error are you talking about?

File: cmd/thriftbreak/main_test.go
Comment (abhinav) [inline cmd/thriftbreak/main_test.go:110]: The message is incorrect. Did you mean "expected an error"?

File: cmd/thriftbreak/main_test.go
Comment (abhinav) [inline cmd/thriftbreak/main_test.go:122]: In Go naming, initialisms are usually all caps (unless they start an unexported variable).
```suggestion
func TestJSONPrinter(t *testing.T) {
```

ref: https://github.com/golang/go/wiki/CodeReviewComments#initialisms

---

Although since this tests both printers, not just JSON, it should probably be named something else.

```suggestion
func TestDiagnosticPrinters(t *testing.T) {
```

File: cmd/thriftbreak/main_test.go
Comment (abhinav) [inline cmd/thriftbreak/main_test.go:131]: Can use backticks here to avoid escaping most of the JSON.

```suggestion
			want:   `{"File":"foo.thrift","Message":"error"}` + "\n",
```

**Review by abhinav — COMMENTED** [review]
Minor comments; should be good to go after them.

File: cmd/thriftbreak/main_test.go
Comment (shirchen) [reply cmd/thriftbreak/main_test.go:107]: https://github.com/kyoh86/scopelint but i'll remove them.

**Pull request merged**
